base/lock.jl

```diff
@@ -164,6 +164,52 @@ function lock(f, l::AbstractLock)
     end
 end
 
+"""
+Lockable(value, lock = ReentrantLock())
+
+Creates a `Lockable` object that wraps `value` and
+associates it with the provided `lock`.
+
+!!! compat "Julia 1.5"
+    Requires at least Julia 1.5.
+"""
+struct Lockable{T, L<:AbstractLock}
+    value::T
+    lock::L
+end
+
+"""
+
+Lockable(value)
+
+Creates a `Lockable` object that wraps `value` and
+associates it with a newly created `ReentrantLock`.
+
+!!! compat "Julia 1.5"
+    Requires at least Julia 1.5.
+"""
+Lockable(value) = Lockable(value, ReentrantLock())
+
+"""
+    lock(f::Function, l::Lockable)
+
+Acquire the lock associated with `l`, execute `f` with the lock held,
+and release the lock when `f` returns. `f` will receive one positional
+argument: the value wrapped by `l`. If the lock is already locked by a
+different task/thread, wait for it to become available.
+
+When this function returns, the `lock` has been released, so the caller should
+not attempt to `unlock` it.
+
+!!! compat "Julia 1.5"
+    Requires at least Julia 1.5.
+"""
+function lock(f, l::Lockable)
+    lock(l.lock) do
+        f(l.value)
+    end
+end
```

Comment:
While this is probably the most useful pattern, it probably makes sense to implement the full `Lock` interface, right? i.e. plain `lock(l::Lockable)`, `unlock(l::Lockable)`, `trylock`, `islocked`, etc.

Reply:
Yeah that would not be difficult.

Should `Lockable` subtype `AbstractLock`? Right now, it does not.

Reply:
I'd say yes.

Reply:
I'd suggest not to. Currently, `lock(f, l::AbstractLock)` works with `f` without an argument:

https://github.com/JuliaLang/julia/blob/ef0e0f2923fbbf6f88d01d9a8f97d62c0c33bdbd/base/lock.jl#L158-L165

I think it'd be confusing the arity of `f` in `lock(f, l::AbstractLock)` depends on the exact `typeof(l)`.

Reply:
Yeah that's a good point.

So I've implemented the rest of the Lock interface. But I did not make `Lockable`a subtype of `AbstractLock`.

Reply:
@quinnj @tkf Are you satisfied that this is resolved?

Reply:
@quinnj @tkf I'm going to mark this as resolved. I have implemented the Lock interface, except for the `trylock` function per @tkf's recommendations.

Reopen this if I am mistaken.